File: python/그래프 탐색/타겟 넘버.py

""" 레벨 2. 타겟 넘버 """

from collections import deque

def solution(numbers, target):
    queue = deque()

    queue.append(numbers[0])
    queue.append(-numbers[0])

    idx = 1

    while idx < len(numbers):

        for _ in range(len(queue)):

            num = queue.popleft()

            queue.append(num + numbers[idx])
            queue.append(num - numbers[idx])

        idx += 1

    return queue.count(target)


## 실행
print(solution([1, 1, 1, 1, 1], 3))     # 5
print(solution([4, 1, 2, 1], 4))        # 2

File: Java/src/그래프탐색/타겟넘버.java
package 그래프탐색;

import java.util.*;

public class 타겟넘버 {
    
    public static void main(String[] args) {

        int[] n1 = {1, 1, 1, 1, 1};
        int t1 = 3;
        System.out.println(solution(n1, t1));

        int[] n2 = {4, 1, 2, 1};
        int t2 = 4;
        System.out.println(solution(n2, t2));

    }

    public static int solution(int[] numbers, int target) {

        Queue<Integer> q = new LinkedList<>();

        q.add(-numbers[0]);
        q.add(numbers[0]);

        for (int i = 1; i < numbers.length; i++) {
            int len = q.size();
            for (int j = 0; j < len; j++) {
                int num = q.poll();
                q.add(num - numbers[i]);
                q.add(num + numbers[i]);
            }
        }

        int answer = 0;
        while (!q.isEmpty()) {
            if (target == q.poll()) {
                answer += 1;
            }
        }
        
        return answer;
    }
}
